homeassistant/components/sunweg/sensor.py

```diff
@@ -0,0 +1,324 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = datetime.timedelta(minutes=5)
+CACHE_INTERVAL = datetime.timedelta(minutes=1)
+
+
+def get_device_list(api: APIHelper, config):
+    """Retrieve the device list for the selected plant."""
+    plant_id = config[CONF_PLANT_ID]
+
+    # Log in to api and fetch first plant if no plant id is defined.
+    login_response = api.authenticate()
+    if not login_response:
+        _LOGGER.error("Username, Password or URL may be incorrect!")
+        return
+    if plant_id == DEFAULT_PLANT_ID:
+        plant_info: list[Plant] = api.listPlants()
+        plant_id = plant_info[0].id
+
+    # Get a list of devices for specified plant to add sensors for.
+    devices = api.plant(plant_id).inverters
+    return [devices, plant_id]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the SunWEG sensor."""
+    config = {**config_entry.data}
+    username = config[CONF_USERNAME]
+    password = config[CONF_PASSWORD]
+    name = config[CONF_NAME]
+
+    # Initialise the library with the username & password
+    api = APIHelper(username, password)
+
+    devices, plant_id = await hass.async_add_executor_job(get_device_list, api, config)
+
+    probe = SunWEGData(api, username, password, plant_id, "total")
+    entities = [
+        SunWEGInverter(
+            probe,
+            name=f"{name} Total",
+            unique_id=f"{plant_id}-{description.key}",
+            description=description,
+        )
+        for description in TOTAL_SENSOR_TYPES
+    ]
+
+    # Add sensors for each device in the specified plant.
+    for device in devices:
+        probe = SunWEGData(api, username, password, device.id, "inverter")
+
+        entities.extend(
+            [
+                SunWEGInverter(
+                    probe,
+                    name=f"{device.name}",
+                    unique_id=f"{device.sn}-{description.key}",
+                    description=description,
+                )
+                for description in INVERTER_SENSOR_TYPES
+            ]
+        )
+
+        for phase in device.phases:
+            phase_probe = SunWEGData(
+                api, username, password, device.id, "phase", phase.name
+            )
+            entities.extend(
+                [
+                    SunWEGInverter(
+                        phase_probe,
+                        name=f"{device.name} {phase.name}",
+                        unique_id=f"{device.sn}-{phase.name}-{description.key}",
+                        description=description,
+                    )
+                    for description in PHASE_SENSOR_TYPES
+                ]
+            )
+
+        for mppt in device.mppts:
+            for string in mppt.strings:
+                string_probe = SunWEGData(
+                    api, username, password, device.id, "string", string.name
+                )
```

Comment:
You shouldn't need to make multiple instances of your data object. Look at the August integration to see how they handle this. They do a similar system instead of a coordinator

Reply:
I'm creating one data for each sensor. Do you suggest one per device.id?

Reply:
I would say best practice is to have one - period, and just have the sensor define how it uses it differently. But I think one per device id would be a good step. 

Again look at August code and I think that will be a good example. 

By just having one you can keep track of authentication, reauthenticate, and not have to authenticate n(sensor number) times. When you make a new object each time you create the sensor you are making a bunch of unneeded api calls.